ninjalooter/raidgroups.py

```diff
@@ -0,0 +1,812 @@
+            self.group_score += player_score
+            available_list.remove(player)
+
+        #
+        # find a tank
+        #
+        target_list = []  # list of (score, Player object) tuples
+        for p in available_list:
+            if p.is_tank():
+                # start with top score
+                player_score = self.MAX_SLOT
+
+                # penalty points for level < 60
+                player_score -= (60 - p.level) * self.LEVEL_PENALTY
+
+                # add to target list
+                target_list.append((player_score, p))
+
+        # sort target list to find highest-scoring targets
+        sorted_target_list = sorted(target_list, key = lambda tuple:tuple[0], reverse = True)
+
+        # add score for top target to group score, and remove target from available_list
+        target_count = len(sorted_target_list)
+        if target_count > 0:
+            (player_score, player) = sorted_target_list.pop(0)
+            self.group_score += player_score
+            available_list.remove(player)
+
+        #
+        # find a slower
+        #
+        target_list.clear()
+        for p in available_list:
+            if p.is_shaman() or p.is_enchanter():
+                # start with top score
+                player_score = self.MAX_SLOT
+
+                # penalty points for level < 60
+                player_score -= (60 - p.level) * self.LEVEL_PENALTY
+
+                # add to list
+                target_list.append((player_score, p))
+
+        # sort target list to find highest-scoring targets
+        sorted_target_list = sorted(target_list, key = lambda tuple:tuple[0], reverse = True)
+
+        # add score for top target to group score, and remove target from available_list
+        target_count = len(sorted_target_list)
+        if target_count > 0:
+            (player_score, player) = sorted_target_list.pop(0)
+            self.group_score += player_score
+            available_list.remove(player)
+
+
+        # add scores for remaining group members - we get at least a few points for having someone in a slot, rather than an empty slot
+        # however, try to avoid duplicates of the classes already added
+        for p in available_list:
+            if (not p.is_priest()) and (not p.is_tank()) and (not p.is_shaman()) and (not p.is_enchanter()):
+                self.group_score += self.MIN_SLOT
+
+        # penalize scores for players in General groups compared to the other more specific groups, to encourage their placement in specific groups
+        self.group_score = round(self.group_score * self.GENERAL_PENALTY)
+
+        # return the sum of the player scores, as matched against the ideal group
+        return self.group_score
+
+
+
+
+
+
+
+
+#
+# class for an EQ raid
+#
+# multiple groups
+# also has the knowledge, given how many total raiders are available, how many groups and what group types / profiles are desired
+#
+class Raid():
```

Comment:
This would also move to models.py / inherit from DictEquals (probably?) ... really it depends on whether this data would EVER be saved out to disk for reloading, I default to "assume we'll save the entire app state" but technically we don't NEED to.